Feature/core movement rework

ninyauwu opened the pull request:
Closes #<issue-number>
Huge movement overhaul. Added sliding, ground pounding, better gun feedback, tightened up movement and made the character first-person.

## Checklist
### 1. Merging
- [x] Did you select the dev branch?
### 2. File Structure
- [x] Are all scripts placed within Assets/Scripts and organized into logical categories using PascalCase (e.g., Core, Player, Level)?
- [x] Are assets like materials, textures and sound files named using lower_snake_case (e.g., diamond_pickaxe.png)?
- [x] Did the changes avoid unnecessary scene modifications? If new changes were needed for testing, were they done in Scenes/Dev/<Specification> and then discarded?
### 3. Code
- [x] Do private fields start with an underscore and use camelCase (e.g., _privateVariable)?
- [x] Are static readonly variables named using UPPER_SNAKE_CASE?
- [x] Are enum types also in UPPER_SNAKE_CASE?
- [x] Are function and variable names clear, concise, and descriptive without using unnecessary abbreviations?
- [x] Are all public functions and fields documented with clear summaries where needed
- [x] Does the code maintain a strict hierarchy, ensuring objects do not fetch references themselves or their parents in the hierarchy?
- [x] Are references passed down through dependency injection as needed?Does each component only depend on its parent in the hierarchy?
- [x] Have unnecessary/testing Debug.Logs been removed?
- [x] Has commented code been removed?
### 4. Logging
- [ ] If possible, did you make a video and did you send it/a link to it in the feature-videos channel?

IcyKoala commented on Project/Assets/Scripts/Player/Movement/PlayerMovementHelper.cs:
```diff
@@ -0,0 +1,136 @@
+    /// <param name="rotation"></param>
+    /// <param name="curInput"></param>
+    /// <param name="dashForce"></param>
+    public static void Dash(ref Vector3 velocity, float rotation, Vector3 curInput, float dashForce)
```
Didn't you remove dash pt2?

ninyauwu replied:
Gone.

IcyKoala commented on Project/Assets/Scripts/Player/Movement/PlayerMovementHelper.cs:
```diff
@@ -0,0 +1,136 @@
+        bool groundContact = false;
+        foreach (ContactPoint contactPoint in collision.contacts)
+        {
+            // Through testing, 0.55 is a good cutoff point where a slope is too steep to reasonably be walked up imo
```
valid, prolly make it a variable then, in case it needs tweaking

ninyauwu replied:
Done.

IcyKoala commented on Project/Assets/Scripts/Player/Movement/SlidingMovementState.cs:
Few different magic numbers here, not sure which ones should be like that

ninyauwu replied:
I separated out the one that might need to be changed.

IcyKoala commented on Project/Assets/Scripts/PlayerController.cs:
```diff
@@ -3,20 +3,27 @@
 using UnityEngine;
 using UnityEngine.InputSystem;
+using UnityEngine.InputSystem.XR.Haptics;
+using static UnityEngine.Rendering.DebugUI;
```
im guessing unity added smthn here? Or did you add smthn for haptics?

ninyauwu replied:
No clue what happened here. These imports were unused. They're gone now.

IcyKoala commented on Project/Assets/Scripts/PlayerController.cs:
```diff
@@ -102,18 +117,25 @@ public void ApplyUpgrade(UpgradeData upgrade)
-                increase = _baseStats.Attack * upgrade.Value;
-                Stats.Attack += increase;
+                increase = PlayerComponentManager.PlayerStats.Attack * upgrade.Value;
+                PlayerComponentManager.PlayerStats.Attack += increase;
```
Feel like *= could be used here to have it be 1 line

ninyauwu replied:
Uhh, how would that work exactly? I'm not seeing it.

IcyKoala replied:
yeah my brain didn't really brain too well with that one, would only work if the upgrade values are 1.[value] rather than 0.[value]

IcyKoala reviewed (changes requested):
- Currently this breaks the bullet-shooting because the player doesn't have the "Explosion Position" of the shotgun set, this needs to be at the front of the shotgun barrel.
- Would be nice if you set the target of the enemies to the character object of the player so the console doesn't get spammed with errors. (All the melee-ranged enemies can be set at once, the shooting enemy needs to be set by itself, i can also easily do this)
- Not sure why you have individual interfaces for sprint and crouch, but also have those functions in the IMovementInputHandler
- not sure why the OnCollision functions have interfaces but im guessing there's a reason
- Why does the MoveState have state-transition handling, shouldn't that be in the relevant states themself? (for example you have a `CurrentState = States["Aerial"];` in MoveState and ` StateSetter.Invoke(Neighbors["Aerial"]);` in GroundedMovementState, which if can have a variety of reasons for existing, but i can't really tell

ninyauwu commented on Project/Assets/ScriptableObjects/Stats/PlayerStats.asset:
```diff
@@ -14,12 +14,20 @@ MonoBehaviour:
-  ModelHeight: 2.4
+  Friction: 15
+  SpeedCap: 60
+  ModelHeight: 4
```
Really this is just two times the amount of distance a raycast is shot down to detect if the player is standing on ground. The name is... not ideal. If you have a better suggestion for a name I'd love to hear it.

IcyKoala replied:
Thats fair, meant more as in general that the character/player pov is higher compared to the old model

ninyauwu commented on Project/Assets/Scripts/Player/Movement/GroundSlammingState.cs:
```diff
@@ -0,0 +1,45 @@
+
+        MovePlayer(ref velocity);
+
+        velocity = new Vector3(velocity.x, velocity.y - (9.81F * GameManager.GravityMultiplier * 3.0F * Time.deltaTime), velocity.z);
```
I don't think there's much point in changing it? If it's much faster you could end up ground pounding through the floor, and if it's much slower you wouldn't notice the acceleration. If you really want me to I can make a stat for it, though.

ninyauwu commented on Project/Assets/ScriptableObjects/Stats/PlayerStats.asset:
```diff
@@ -14,12 +14,20 @@ MonoBehaviour:
-  TurnTime: 0.1
+  SlidingAcceleration: 2.6
+  JumpForce: 5
+  DashForce: 20
```
Uh, I thought I did. Good catch, it's gone now.

ninyauwu commented on Project/Assets/Scripts/Input/UnityInputEvent.cs:
This class is unused. I got rid of it.
I was testing Unity events but I ended up getting rid of it in favor of Matheus' solution.

ninyauwu commented on Project/Assets/Scripts/Player/Movement/GroundedMovementState.cs:
```diff
@@ -0,0 +1,96 @@
+        {
+            PlayerMovementHelper.RememberedHeldSlide = false;
+            //PlayerMovementHelper.ModifyByGroundSlope(ref velocity);
+            //PlayerRigidbody.velocity = velocity * 3.0F;
```
Done.

ninyauwu commented:
> Currently this breaks the bullet-shooting because the player doesn't have the "Explosion Position" of the shotgun set, this needs to be at the front of the shotgun barrel.
Fixed.
> Would be nice if you set the target of the enemies to the character object of the player so the console doesn't get spammed with errors. (All the melee-ranged enemies can be set at once, the shooting enemy needs to be set by itself, i can also easily do this)
Fixed.
> Not sure why you have individual interfaces for sprint and crouch, but also have those functions in the IMovementInputHandler
Removed deprecated interfaces.
> not sure why the OnCollision functions have interfaces but im guessing there's a reason
Because collision is detected by PlayerController, but handled by the current state of the controller. So it's to delegate these functions.
> Why does the MoveState have state-transition handling, shouldn't that be in the relevant states themself? (for example you have a CurrentState = States["Aerial"]; in MoveState and  StateSetter.Invoke(Neighbors["Aerial"]); in GroundedMovementState, which if can have a variety of reasons for existing, but i can't really tell
Why would it need to be done by the individual states? OnAttackInput is not really handled by the substates but, by the shotgun, so no need. OnCrouchInput is a bit of an edge case, but I would have to write the entire grounded check in both GroundedMovementState and AerialMovementState so I didn't bother.
